In [10]:
#Import packages
from scipy.optimize import minimize
import numpy as np

In [11]:
def func(x):
    
    return x**2

result = minimize(func, 1)
x_opi = result.x

print(x_opi)

[-7.45036449e-09]


In [12]:
def func(x):
    
    return x[0]**2 + x[1]**2

result = minimize(func, np.array([1,1]))
x_opi = result.x

print(x_opi)

[-1.07505143e-08 -1.07505143e-08]


In [13]:
#This is the normal inverse gamma function
def norminvgamma_pdf(x, sigma2, mu, nu, alpha, beta):
    '''
    The probability density function of the normal-inverse-gamma distribution at
    x (mean) and sigma2 (variance).
    '''
    return (
        sts.norm.pdf(x, loc=mu, scale=np.sqrt(sigma2 / nu)) *
        sts.invgamma.pdf(sigma2, a=alpha, scale=beta))

In [18]:
#Method 1

def param_func(x): #(mu_0, nu_0, alpa_0, beta_0)
    
    #first term is for the expected value for the mean mean
    #second term is for the expected value for the variance beta/(alpha-1
    #the third term is the variance for the for the mean beta/((alpa-1)*nu)
    #the forth term is the variance for the variance beta^2/((alpha-1)^2*(alpha-2))
    return (x[0]-2.3)**2 + (x[3]/(x[2]-1)-2.75)**2 + (x[3]/((x[2]-1)*x[1])-0.5)**2 + (x[3]**2/((x[2]-1)**2*(x[2]-2))-1)**2


result = minimize(param_func, [2,1,3,2])

mu_0 = result.x[0]
nu_0 = result.x[1]
alpha_0 = result.x[2]
beta_0 = result.x[3]

print(f"mu = {mu_0}")
print(f"nu = {nu_0}")
print(f"alpha = {alpha_0}")
print(f"beta = {beta_0}")

mu = 2.3000000229838147
nu = 5.4999906215577
alpha = 9.562498590173872
beta = 23.546877384848045


## Method 2

mean = 2.3

Variance = 2.75 = beta/(alpha-1)

Var[mean] = 0.5 = beta/((alpha-1)*nu) = 0.5 * 2.75/nu --> nu = 5.5

Var[variance] = 1 = beta^2/((alpha-1)^2*(alpha-2)) = variance^2/(alpha-2) = 2.75^2/(alpha-2) --> alpha = 2.75^2+2 = 9.56

Variance = 2.75 = beta/(9.56-1) --> beta = 2.75*8.56 = 23.54

mu_0 = 2.3
nu_0 = result.x[1]
alpha_0 = result.x[2]
beta_0 = result.x[3]